core/src/main/java/org/infinispan/expiration/impl/ExpirationManagerImpl.java

```diff
@@ -143,9 +146,14 @@ public boolean isEnabled() {
    }
 
    @Override
-   public CompletableFuture<Boolean> entryExpiredInMemoryFromIteration(InternalCacheEntry<K, V> entry, long currentTime) {
-      // Local we just remove the entry as we see them
-      return entryExpiredInMemory(entry, currentTime, false);
+   public boolean entryExpiredInMemoryFromIteration(InternalCacheEntry<K, V> entry, long currentTime) {
+      // If entry was expired and we have store this can block - so fire in separate thread to remove the entry
+      if (persistenceManager.isEnabled()) {
+         blockingExecutor.submit(() -> entryExpiredInMemory(entry, currentTime, false));
```

Comment:
1. Can we check if there are actual stores active and not just loaders?
2. `persistenceManager.deleteFromAllStores()` already does the work on another thread, couldn't we remove the `join()` from `deleteFromStoresAndNotify()` instead?

Reply:
We can't just remove the join. The problem is the fact that the the store removal is done inside the DataContainer compute method - which doesn't allow us to return anything as it needs to be completed before returning. If not we lose consistency guarantees.

We would have to make a new method that does the removal sync in the compute block. This should probably be done though to make sure we are using less threads.

Reply:
In regards to making sure there are only stores, we could make a new method on PersistenceManager I guess that says if there is a store or not.

Reply:
> which doesn't allow us to return anything as it needs to be completed before returning.

Does it really need to be completed before returning? If we return from `compute` and the entry still exists in the store, the timestamp in the store is still older and will be expired when another command reads it. IMO the only problem is a thundering herd effect when multiple commands try to read the expired key, they all go to the store, and then they all try to remove it from the store.

An alternative would be to add a `CompletableFuture` field to all the entries and only use `compute()` to change the future atomically, similar to `RemoteTransaction.enterSynchronizationAsync()`.

Reply:
I am not a fan either way. To be honest this issue should be taken care of in the JIRA I created for it a while back https://issues.jboss.org/browse/ISPN-10360, not here.

Reply:
Ok, if there's a JIRA for it already then I'll integrate this as is. It was the extra thread from `deleteFromAllStores` that got my attention anyway, and you've dealt with that.